Websocket client api

cjswedes opened the pull request.

cjswedes committed:
fixup tests

cjswedes committed:
start fleshing out client receive loop and close

cjswedes committed:
rename ServerWebSocket to WebSocketServer

Commit:
fill in default sends

Co-authored-by: Carter Swedal <cjswedes@users.noreply.github.com>

Commit:
add keep_alive config

Co-authored-by: Carter Swedal <cjswedes@users.noreply.github.com>

Commit:
Make masking easier

Co-authored-by: Carter Swedal <cjswedes@users.noreply.github.com>

cjswedes committed:
Failure to receive pong causes connection close

cjswedes committed:
attempt to cleanup some tests, some were fixed.

cjswedes committed:
Do we want a WebSocket, a WebSocketClient, and a WebSocketServer object?

cjswedes committed:
Add message and send(message) api to WebSocketClient

cjswedes committed:
fixup! Add message and send(message) api to WebSocketClient

cjswedes committed:
tried to run a test, but I am failing to get the environment set up right

cjswedes committed:
Adding cosock to the rockspec did not work to get the test going either.

I tried a bunch of different things:

**Reset the folder:**

`rm -rf ./.luarocks && rm -rf ./lua_modules && rm -f
./lua && rm -f ./luarocks`

`luarocks init`
`luarocks make`

**Add more search areas to the LUA_PATH**
`export
LUA_PATH="./lua_modules/share/lua/5.3/?.lua;./lua_modules/share/lua/5.3/?/init.lua;./?.lua;./?/init.lua;;"`

**I still get errors when pulling cosock from client.lua

cjswedes committed:
I got one client test running finally!!

Not sure if the rockpec file changes are necessary or not

cjswedes committed:
make masking more intuitive for the user

cjswedes committed:
attempt at figuring out the handshake tests. Something is still not working with the headers and the mock requests...

cjswedes committed:
more sending tests, need to be able to expect send on the mock socket still

cjswedes committed:
I tried to finalize the api by annotating it.

I left some more todos in client.lua, but then moved into writing a
script that would allow me to test the basic api functionality with
an echo server.

cjswedes committed:
fixup! I tried to finalize the api by annotating it.

cjswedes commented:
@FreeMasen Can you review the API annotations in `client.lua` to see if there is anything that needs to change wrt the websocket client api parameters, and return values?

FreeMasen commented on lustre/client.lua:
```diff
@@ -1,30 +1,106 @@
         config = config or Config.default(),
+        _tx = _tx,
+        _rx = _rx,
+        message_cb = message_cb,
+        error_cb = error_cb, 
+        close_cb = close_cb
```
These would always be `nil`, if you wanted to get them from the `...` parameter you could do that by wrapping it in a table and then indexing those `extra_args = {...}` and then `message_cb = extra_args[1]`, though it might be nice to implement the builder pattern here, so the `register_*` functions would return `self` this would allow 

```lua
local ws = WebSocketClient.new(sock, '', cfg):register_message_handler(print):register_error_handler(print)
```

FreeMasen reviewed (commented):
Overall I think it looks pretty good with just one suggestion on the constructor arguments.

One thing I keep meaning to explore is how different we would expect the client and server api's to actually be. In my head they seem like they are similar enough to be able to export a single `WebSocket` with a `client` and `server` constructor that setups up the config to work as expected. This isn't super important right now as we could magic ourselves into a solution later but thought it was worth mentioning.

cjswedes committed:
Fixup api

cjswedes committed:
Debugging the handshake, and making progress

We are sending the handshake request, but are failing to validate
response.

cjswedes committed:
Handshake with echo server is successful

cjswedes committed:
send happy path working
receive echo from echo server working

How do we want the receive api to work? and frankly, the send api as
well...We were headed in the direction of having a run loop that
executes callbacks for messages and errors. We could integrate the send
into that also. Or we could just expect the user to call receive when
they expect something to come in. I think the run loop route was taken
because of bi directional quality of websockets, but I think both can make
sense.

cjswedes committed:
fixup! send happy path working receive echo from echo server working

cjswedes commented:
@FreeMasen I am curious about the `receive()` API? I am just starting to debug that API with the echo server, and it seems like the way I use it in [`test_client_script`](https://github.com/cosock/lustre/blob/07a5a1d3a4b8ed4db34403fd5e0a40d4bee68f80/spec/test_client_script.lua#L44-L66) does not really make sense. That API usage seems to contradict the run loop functionality that we originally implemented to handle incoming messages, and control flow messages with callbacks. I was wondering if you had any suggestions for what usage would look like after we have successfully `connect`ed to a websocket serve.

FreeMasen commented:
I would agree that `receive()` doesn't really make sense, since a web socket message can come it at any time. Instead the run loop would be handing all complete messages to the [function passed to `register_message_cb`](https://github.com/cosock/lustre/blob/07a5a1d3a4b8ed4db34403fd5e0a40d4bee68f80/spec/test_client_script.lua#L33-L35)

I am thinking this is what the test should look like: [gist](https://gist.github.com/FreeMasen/5f6a8c84393fc4ccaf9a2998b616a220).

Commit:
minor tweeks for receive

cjswedes committed:
Merge pull request #3 from cosock/spawn-receive-loop

minor tweeks for receive

cjswedes committed:
add a protocol to the connection

cjswedes committed:
Move tcp socket connection into connect api

cjswedes committed:
Running test cases against the autobahn framework

cjswedes committed:
Most text sending and ping tests passing WIP

cjswedes committed:
fixup! Most text sending and ping tests passing WIP

cjswedes committed:
Add support for binary messages

cjswedes committed:
cleanup some print statements

cjswedes committed:
opcode and rsv bit test cases

cjswedes committed:
Fragmentation receive first test passing

cjswedes committed:
fragmentation tests passing

cjswedes committed:
cleanup a little

cjswedes committed:
closing tests are passing

cjswedes committed:
add timeout case for receiving on socket

FreeMasen commented on lustre/frame/init.lua:
```diff
@@ -68,20 +102,37 @@ function Frame:payload_len()
+end
+
+local function seed_once()
+  if seeded then return end
```
Am I missing where `seeded` is defined? I feel like we might want to have this be a file level local variable, unless I am missing something

FreeMasen commented on lustre/message.lua:
```diff
@@ -0,0 +1,21 @@
+---@field protocols string[] List of requested protocols
+---@field extensions table[] List of requested extensions
+---@field key string|nil Signing key
+---@field accept string|nil Sec-WebSocket-Accept header value
```
These don't seem to be correct

FreeMasen commented on lustre/ws.lua:
```diff
@@ -0,0 +1,351 @@
+---@param message Message
+---@return err string|nil
+function WebSocket:send(message)
+    return "not implemented"
```
```suggestion
    return nil, "not implemented"
```

FreeMasen commented on lustre/ws.lua:
```diff
@@ -0,0 +1,351 @@
+    local close_frame = Frame.close(close_code, reason):set_mask()
+    local sent, err = self.socket:send(close_frame:encode())
+    if not sent then return nil, err end
+    print("SENT CLOSE FRAME: \n", utils.table_string(close_frame), "\n\n")
```
This should probably be using `log.debug` over `print` which would require including `logface` in the dependencies

FreeMasen commented on lustre/ws.lua:
```diff
@@ -0,0 +1,351 @@
+                    return
+                elseif err == "invalid opcode" or err == "invalid rsv bit" then
+                    self:close(CloseCode.protocol(), err)
+                elseif err == "timeout" and self.error_cb then
```
We may want to add a todo here since timeout means we should be re-trying to complete the pending frame. I think treating it like an error for now is ok. We should probably include the partial frame in the callback arguments

FreeMasen commented on lustre/ws.lua:
```diff
@@ -0,0 +1,351 @@
+                if not frame:is_final() then
+                    multiframe_message = true
+                end
+            elseif frame.header.opcode.sub   == 'continue' and not multiframe_message then
```
```suggestion
            elseif frame.header.opcode.sub == 'continue' and not multiframe_message then
```

FreeMasen commented on lustre/ws.lua:
```diff
@@ -0,0 +1,351 @@
+            --aggregate payloads
+            if not frame:is_final() then --todo and we are continuing something
+                received_bytes = received_bytes + frame:payload_len()
+                --Dont build up a message payload that is bigger than max message size
+                if received_bytes <= self.config.max_message_size then
+                    table.insert(partial_frames, frame.payload)
+                end
```
Probably want a todo here since we are truncating the total payload when it exceeds the maximum, deciding what we actually want to do might be a bigger conversation

FreeMasen commented on lustre/ws.lua:
```diff
@@ -0,0 +1,351 @@
+            end
+        elseif recv[1] == self._rx then
+            self._rx:recieve()
+            --TODO @FreeMasen what events were you thinking would be injected into 
```
I was expecting this task would handle all of the socket communications. So the two `send` methods above would send on this channel and this task would then handle the `socket:send`, most likely with a retry on timeout, and sending in multiple parts if needed.

cjswedes replied:
I think I have made this asychronous so that th erun loop handles the socket operations. However, the retry on timeout and the testing of the fragmentation has yet to be completed.

FreeMasen commented on spec/utils.lua:
```diff
@@ -29,12 +117,19 @@ local function table_string(v, pre, visited)
     end
     return string.format('%s\n%s}', ret, orig_pre)
 end
+--]]
```
This big comment should probably be cleaned up

FreeMasen commented on spec/utils.lua:
```diff
@@ -94,4 +189,5 @@ return {
-}
\ No newline at end of file
+    table_string = table_string
+}
```
nit: final new line

FreeMasen commented on spec/ws-client-hello.py:
```diff
@@ -0,0 +1,19 @@
+        greeting = await websocket.recv()
+        print(f"< {greeting}")
+
+asyncio.get_event_loop().run_until_complete(hello())
```
nit: final new line

FreeMasen commented on spec/frame_spec.lua:
```diff
@@ -5,54 +5,54 @@ local OpCode = require 'lustre.frame.opcode'
+      utils.assert_eq(f.payload, CloseCode.normal():encode())
     end)
+  end)
 end)
```
nit: final new line

FreeMasen commented on spec/utils.lua:
```diff
@@ -1,4 +1,91 @@
+        local v = val[k]
+        previously_printed[val] = name
+        if #val > 0 and type(k) == "number" then
+          tabStr =  tabStr .. stringify_table_helper(v, nil, multi_line, indent + 2, previously_printed) .. ", " .. multi_line_str
+        else
+          tabStr =  tabStr .. stringify_table_helper(v, k, multi_line, indent + 2, previously_printed) .. ", ".. multi_line_str
```
nit: super long lines

FreeMasen reviewed (approved):
Overall this looks very good, just a few small things. We should probably run a lua formatter across the project, if nothing else to align the string quote character. [You could use the luxure config](https://github.com/cosock/luxure/blob/master/.luafmt)

cjswedes committed:
simple review fixups and todos

cjswedes committed:
fixup! use log.debug instead of print statements

cjswedes committed:
make send_text, send_bytes, and close apis send asynchrounously

cjswedes committed:
Add lua format and format source

cjswedes commented:
Gonna merge this since I think its at an acceptable spot. When we end up merging the `next` branch into master we can handle creating a new luarock spec and ensuring all tests we want to have passing are passing. The tests that are failing now seem to be related to the weird `cosock.socket` receive functionality that is already encapsulated in an [issue](https://github.com/cosock/cosock/issues/17).
